### PR TITLE
feat(analyze/js): implement `noShadow`

### LINKED ISSUE #5346: 📎 Port `no-shadow` from eslint
### Description

Port the [`no-shadow`](https://eslint.org/docs/latest/rules/no-shadow) rule from eslint. We already have `noShadowRestrictedNames`, which can be used as a reference for implementing this rule.

Suggested name: `noShadow`

**Want to contribute?** Lets you know you are interested! We will assign you to the issue to prevent several people to work on the same issue. Don't worry, we can unassign you later if you are no longer interested in the issue! Read our [contributing guide](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md) and [analyzer contributing guide](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by dyc3**
## Summary

This PR implements `noShadow`, which is a port of eslint's `no-shadow` rule.

It doesn't include any options for the time being (as usual for new rules).

closes #5346

## Test Plan

Added some snapshot tests. I also briefly went through the eslint rule to pick up some edge cases that I hadn't thought about.

### CODE REVIEW

**Comment by dyc3** [comment]
Oof, quite the noticeable perf regression. Somewhat expected because my implementation is rather naive.

File: crates/biome_js_analyze/src/lint/nursery/no_shadow.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    /// ```js
+    /// const foo = "bar";
+    /// if (true) {
+    ///    const qux = "baz";
+    /// }
+    /// ```
+    ///
+    pub NoShadow {
+        version: "next",
+        name: "noShadow",
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/no_shadow.rs:58]: What about `noShadowedVariables`? The other name is very generic

File: crates/biome_js_analyze/src/lint/suspicious/no_shadow_restricted_names.rs [hunk 2]
```diff
@@ -10,6 +10,8 @@ use biome_rowan::AstNode;
 declare_lint_rule! {
     /// Disallow identifiers from shadowing restricted names.
     ///
+    /// See also: `noShadow`
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/suspicious/no_shadow_restricted_names.rs:13]: Link missing

File: crates/biome_js_analyze/tests/specs/nursery/noShadow/invalid.js
Comment (ematipico) [inline crates/biome_js_analyze/tests/specs/nursery/noShadow/invalid.js:1]: Eslint has a very extensive testing suite. Should we borrow some? https://github.com/eslint/eslint/blob/main/tests/lib/rules/no-shadow.js

File: .changeset/twenty-meals-reply.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@biomejs/biome": minor
+---
+
+Added new lint rule `noShadow`, a port of eslint's `no-shadow`. Contributed by @dyc3.
```
Comment (ematipico) [inline .changeset/twenty-meals-reply.md:5]: Couple of notes:
- we usually add a link to the new rule 
- we should explain what we rule does with a small example 
- we don't need the "contributed by" anymore, the changesets does it for us

File: crates/biome_js_analyze/tests/specs/nursery/noShadow/invalid.js
Comment (dyc3) [reply crates/biome_js_analyze/tests/specs/nursery/noShadow/invalid.js:1]: Yes I was looking for that but couldn't find it. Thanks!

File: crates/biome_js_analyze/src/lint/nursery/no_shadow.rs (hunk 1 above)
Comment (dyc3) [reply crates/biome_js_analyze/src/lint/nursery/no_shadow.rs:58]: From an end user's perspective, the rule doesn't just cover "variables". It covers all bindings, so functions, classes, imports, etc. I mean yes, technically they _are_ all variables in that you can read/write to them as such, but you wouldn't refer to them like that.

Plus, this is also a rare case where we can match the eslint rule name. :)

File: crates/biome_js_analyze/src/lint/nursery/no_shadow.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        language: "js",
+        recommended: false,
+        sources: &[
+            RuleSource::Eslint("no-shadow"),
+            // uncomment when we can handle the test cases from typescript-eslint
+            // RuleSource::EslintTypeScript("no-shadow"),
+        ],
```
Comment (dyc3) [inline crates/biome_js_analyze/src/lint/nursery/no_shadow.rs:68]: I have excluded the unit tests from `typescript-eslint`, and therefore the rule source here, because of #5822.

File: crates/biome_js_analyze/src/lint/nursery/no_shadow.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                // redeclarations are not shadowing, they get caught by `noRedeclare`
+                continue;
+            }
+            if is_declaration(&binding)
+                && is_declaration(&upper_binding)
+                && upper_binding.syntax().text_range().start() >= binding.scope().range().end()
+            {
+                // the shadowed binding must be declared before the shadowing one
+                continue;
+            }
```
Comment (dyc3) [inline crates/biome_js_analyze/src/lint/nursery/no_shadow.rs:146]: This check is probably the most dubious one here because I don't have a source for it. I'm not entirely sure which part of the JS spec would be relevant.

**Comment by dyc3** [comment]
I'm not sure why the windows test job is failing. Doesn't feel related to this PR.

**Review by ematipico — APPROVED** [review]
Awesome work! 🚀  I think we can merge it as first phase of the implementation until we fix the sematic model for TS types

**Comment by dyc3** [comment]
I think its fine to close the issue. We can just leave the rule in nursery until the TS cases are fixed.
